android/src/toga_android/widgets/canvas.py

```diff
@@ -236,14 +237,20 @@ def _text_paint(self, font):
         paint.setTextSize(self.scale_out(font.size()))
         return paint
 
+    # This has been separated out, so that it can be mocked during testing.
+    def _native_get_background(self):
+        return self.native.getBackground()
```

Comment:
This isn't an acceptable approach. Mocking is a *last resort* for testing. It's used in a handful of places related to permissions and hardware APIs because using those APIs requires passing control to a separate process, at which point the test suite loses the ability to execute. Background color is an entirely internal mechanism, and isn't encumbered by this restriction.

It may be *difficult* to interrogate - but that doesn't mean we can mock it, because we're trying to verify the actual behavior.

Reply:
During the test, coverage reported missing coverage for:
https://github.com/beeware/toga/blob/dad68bb0fd519537778bb0401e42ad0ee8c3a365/android/src/toga_android/widgets/canvas.py#L245-L247

```
I/python.stdout: Name                                                                                                        Stmts   Miss Branch BrPart  Cover   Missing
I/python.stdout: -------------------------------------------------------------------------------------------------------------------------------------------------------
I/python.stdout: data/data/org.beeware.toga.testbed/files/chaquopy/AssetFinder/requirements/toga_android/widgets/canvas.py     153      0     28      1  99.4%   245->247
I/python.stdout: -------------------------------------------------------------------------------------------------------------------------------------------------------
I/python.stdout: TOTAL                                                                                                        2190      0    324      1  99.9%
```
So, 245->247 means it is reporting missing coverage for the missing else branch. The else branch will be triggered when `self.native.getBackground()` returns None. Since, it is an internal platform method, I cannot make it return None without mocking it. The other way was to ignore the missing else branch with `#pragma: no branch`, which you have told me to avoid. So, is there any other way to test the missing else branch or am I missing something?

Reply:
NVM, I just realized there was a much simpler way to test this. Thanks :) So, the question remains, should I move the test to testbed or keep it in the probe, since it is implementation specific?

Reply:
There might be a specific edge case that is needed to exercise that test, but if the other backends don't have an equivalent "if background" branch, then they will run the code without branching, giving double coverage of one specific code path. That's not a problem - it's OK to test something more than once on iOS (et al) if it means we get 100% coverage on Android as well; if nothing else, it's an indication that there *are* two different test configurations, and we need to test both, even if the "default" implementation works for both cases on iOS. 

However, I'm confused because that line *did* have coverage prior to this change, and it has coverage on every other platform. This, to me, indicates that there's a bigger problem here. Either this branch of code was covering over an edge case that no longer exists, or there's a case that isn't being tested.